crates/flux-refineck/src/checker.rs

```diff
@@ -263,6 +307,74 @@ impl<'ck, 'genv, 'tcx, M: Mode> Checker<'ck, 'genv, 'tcx, M> {
             ck.check_basic_block(infcx, env, bb)?;
         }
 
+        ck.check_trait_impl_subtyping(infcx, def_id, span)?;
+
+        Ok(())
+    }
+
+    fn has_trusted_impl(genv: &GlobalEnv, def_id: DefId) -> bool {
+        if let Some(did) = genv
+            .resolve_id(def_id)
+            .as_maybe_extern()
+            .map(|id| id.local_id())
+        {
+            genv.trusted_impl(did)
+        } else {
+            false
+        }
+    }
+
+    /// Trait subtyping check, which makes sure that the type for an impl method (def_id)
+    /// is a subtype of the corresponding trait method.
+    fn check_trait_impl_subtyping(
+        &mut self,
+        mut infcx: InferCtxt<'_, 'genv, 'tcx>,
+        def_id: LocalDefId,
+        span: Span,
+    ) -> Result {
+        let genv = infcx.genv;
+
+        if let Some((trait_ref, trait_method_id)) =
+            find_trait_item(&genv, def_id).with_span(span)?
+        {
+            // Skip the check if either the trait-method or the impl-method are marked as `trusted_impl`
+            if Self::has_trusted_impl(&genv, trait_method_id)
+                || Self::has_trusted_impl(&genv, def_id.to_def_id())
+            {
+                return Ok(());
+            }
+
+            let trait_fn_sig = genv.fn_sig(trait_method_id).with_span(span)?;
+            let tcx = genv.tcx();
+            let impl_id = tcx.impl_of_method(def_id.to_def_id()).unwrap();
+            let impl_args = GenericArg::identity_for_item(genv, def_id).with_span(span)?;
+            let trait_to_impl_args = &trait_ref.args;
+            let trait_args = impl_args.rebase_onto(&tcx, impl_id, trait_to_impl_args);
+            let trait_refine_args =
+                RefineArgs::identity_for_item(&genv, trait_method_id, |i, param| {
+                    if param.mode == rty::InferMode::KVar {
+                        // See `tests/pos/surface/trait13.rs` for an example
+                        // we need to "define_var" the KVar in the trait's super-type
+                        // TODO: maybe the `trait_refine_args` business can be done inside the `check_fn_subtyping`?
+                        infcx.define_vars(&param.sort)
```

Comment:
I find this extremely confusing and I think it's wrong in subtle ways. It'd probably help to refactor this outside of `Checker` and use a separate `InferCtxt` (there's no reason to make this part of the constraint generated for the body). Moving this out of `Checker` could also help with improving the error message